sdk/core/core-rest-pipeline/src/nodeHttpClient.ts

```diff
@@ -244,28 +250,35 @@ class NodeHttpClient implements HttpClient {
   }
 
   private getOrCreateAgent(request: PipelineRequest, isInsecure: boolean): http.Agent {
-    if (!request.disableKeepAlive) {
-      if (isInsecure) {
-        if (!this.httpKeepAliveAgent) {
-          this.httpKeepAliveAgent = new http.Agent({
-            keepAlive: true,
-          });
-        }
+    if (!request.tlsSettings && request.disableKeepAlive) {
+      return isInsecure ? http.globalAgent : https.globalAgent;
+    }
+    const settingsIndex = request.tlsSettings ?? {};
+    let agent: http.Agent | undefined = isInsecure
+      ? this.httpKeepAliveAgent?.agent
+      : this.httpsKeepAliveAgent?.get(settingsIndex)?.agent;
 
-        return this.httpKeepAliveAgent;
-      } else {
-        if (!this.httpsKeepAliveAgent) {
-          this.httpsKeepAliveAgent = new https.Agent({
-            keepAlive: true,
-          });
-        }
+    if (agent) {
+      return agent;
+    }
 
-        return this.httpsKeepAliveAgent;
-      }
-    } else if (isInsecure) {
-      return http.globalAgent;
+    const agentOptions: https.AgentOptions = {
+      ...request.tlsSettings,
+      keepAlive: !request.disableKeepAlive,
+    };
+
+    if (isInsecure) {
+      agent = new http.Agent(agentOptions);
+      this.httpKeepAliveAgent = { agent };
+      return agent;
     } else {
-      return https.globalAgent;
+      agent = new https.Agent(agentOptions);
+      this.httpsKeepAliveAgent.set(settingsIndex, {
+        agent,
+        tlsSettings: request.tlsSettings,
```

Comment:
do we need this? Isn't `request.tlsSettings === settingsIndex` whenever it's defined, so if you have a reference to `settingsIndex` you already have the reference to the tlsSettings?

Reply:
Okay so just assign settingsIndex instead of request.tlsSettings right?

Reply:
Or just remove this property from the cache object altogether, you can't retrieve it without already having the settings object

Reply:
WeakMaps are not iterable